File: backend/src/main/java/com/commercesuite/common/util/Money.java
package com.commercesuite.common.util;

/**
 * MONEY_SPEC.md: integer paise, ISO-4217 INR only.
 * Never use double/float for money anywhere in the codebase.
 */
public final class Money {
    public static final String CURRENCY = "INR";
    private Money() {}
    public static long rupeesToPaise(long rupees) { return Math.multiplyExact(rupees, 100L); }
    public static long paiseFromMajorMinor(long rupees, int paise) {
        if (paise < 0 || paise > 99) throw new IllegalArgumentException("paise out of range");
        return Math.addExact(Math.multiplyExact(rupees, 100L), paise);
    }
    public static String format(long paise) {
        long abs = Math.abs(paise);
        return (paise < 0 ? "-" : "") + "₹" + (abs / 100) + "." + String.format("%02d", abs % 100);
    }
}

File: src/lib/money.ts
/**
 * MONEY_SPEC bridge. Backend stores money as integer paise (BigDecimal-safe).
 * Frontend historically uses `number` (rupees) — these helpers are the ONE
 * place that crosses the boundary so no floating-point drift leaks in.
 */
export function paiseToRupees(paise: number | string | null | undefined): number {
  if (paise == null) return 0;
  const n = typeof paise === "string" ? Number(paise) : paise;
  if (!Number.isFinite(n)) return 0;
  return Math.round(n) / 100;
}

export function rupeesToPaise(rupees: number | string | null | undefined): number {
  if (rupees == null) return 0;
  const n = typeof rupees === "string" ? Number(rupees) : rupees;
  if (!Number.isFinite(n)) return 0;
  // Avoid 0.1 + 0.2 drift: format then strip.
  return Math.round(n * 100);
}

export function formatInr(paise: number | string | null | undefined): string {
  const rupees = paiseToRupees(paise);
  return new Intl.NumberFormat("en-IN", {
    style: "currency",
    currency: "INR",
    maximumFractionDigits: 2,
  }).format(rupees);
}